Feat/ddrtv 29 auth0 backoffice tutorial (#620)
mia-platform/documentation: docs/getting-started/tutorials/backoffice/auth0-integration.mdx

* The [Backoffice Tutorial Basics](/getting-started/tutorials/backoffice/basics.mdx).

This tutorial will use the following application:
* [Secure API Gateway](../../../runtime_suite_applications/secure-apigateway/overview)

and some of the concepts described on this tutorial:
* [Dev Portal - Authentication Configuration](../../../runtime_suite_applications/dev_portal/authentication_configuration)

## Setup the Microservices

You can install almost all the microservices for this tutorial using the [Secure API Gateway](../../../runtime_suite_applications/secure-apigateway/overview) application:
1. Go to Applications and click on the `Create New Application` button.
2. Search for `Secure ApiGateway` and click on the card.
![Secure ApiGateway Application Marketplace](img/auth0_01-secure-api-gateway-application.png)
3. Follow the wizard, selecting if you want to install new microservices or use existing ones.
4. At the end, click the `Create` button and then save the new Application.

This application will create the following:

> * Microservices 

Reviewer: I would say somehow that these are MIa-Platform plugins. They do not require the user to write code but just the configuration

Author: I've changed the sentence under the **Configure the Microservices and Auth0** section with the following:

> All the microservices that you have added in the previous step are Mia-Platform plugins, so you don't need to write code, but just to configure them.

        "DELETED"
      ],
      "visualizationOptions": {
        "hidden": true
      },
      "formOptions": {
        "hidden": true
      }
    },
    "userGroup": {
      "type": "string",
      "label": "Gruppo"
    },
    "crudSchema": {
      "type": "object",
      "label": "Schema CRUD",
      "default": {
        "type": "object",
        "required": [
          "email",
          "userGroup",
          "name"
        ],
        "properties": {
          "email": {
            "type": "string",
            "format": "email"
          },
          "userGroup": {
            "type": "string"
          },
          "name": {
            "type": "string"
          }
        }
      },
      "visualizationOptions": {
        "hidden": true
      },
      "formOptions": {
        "hidden": true
      }
    },
    "label": {
      "label": "Etichetta",
      "type": "string"
    },
    "authUserCreationDisabled": {
      "label": "Disabilita creazione utente",
      "type": "boolean",
      "default": false,
      "visualizationOptions": {
        "hidden": true
      },
      "formOptions": {
        "hidden": true
      }
    }
  },
  "required": [
    "name"
  ]
}
```

> Then you can add the newly created pages to the Backoffice homepage:

Reviewer: ```suggestion
Then you can add the newly created pages to the Backoffice menu:
```

6. Click on the `Save` button.
7. Repeat the first 6 steps to add the `user-manager-service` configurations page, choosing your favourite *Icon* and with the following changes:
```
Id: ums-configurations
Label: Configurations
```

Then you have to set the user menu:
1. Click on the `Layout` tab.
2. Select `Layout` from the left-side menu.
3. Click on the `Edit property` button under the `Login and user context menu` section on the right-side menu, under the `Advanced Properties`.
4. On the opened modal click fill the `User Context URL` with the value `/userinfo`
5. Click on the `Add property` button related to `Logout` section.
6. Fill the `Redirect Url` with `/logout?redirect=/web-login?redirect=/backoffice`.
7. Click on the `Save` button.

## Configure the authentication process

In order to protect your Backoffice you need to follow these steps:
1. Go to Endpoints, select `/backoffice` and check the `Authentication required` checkbox under the `Endpoint settings`
2. Set the following variables on `Authentication Service`:
  - `USERINFO_URL`: http://user-manager-service/userinfo
  - `AUTHORIZATION_HEADERS_TO_PROXY`: cookie,authorization,client-type

> The final configuration step is to change the API key client type with the one defined as defaultClient in the Auth0 Client configuration.

Reviewer: how can I do this step?

![Backoffice Homepage](img/auth0_07-homepage.png)

## Create a user using Backoffice
In order to create your first user you need to create a configuration. To do that:
1. Click on `Configurations` page.
2. Click on `Add new` button.
3. Fill the form with the following parameters:
![Backoffice Page Composer](img/auth0_08-create-config.png)
4. Click on `Create` button and `Confirm` on the warning modal.

Now it's possible to create the first user:
1. Click on `Users` page.
2. Click on `Add new` button.
3. Fill the form with the `Name` and `Email` of your first user and using `user` as `Group`.
4. Click on `Create` button and `Confirm` on the warning modal.

Congratulations! You have successfully created your first user using the backoffice.
If you try to access the Backoffice with this new user you won't see the users management pages that are protected with the ACLs created in the previous steps.
Now if you delete the created user it will be blocked on Auth0 and marked as `TRASH` inside the CRUD collection.

> ## Backoffice Tutorials

Reviewer: I would add a sectoin where we actually protect some pages/component using ACLs

Author: I've added a section inside **Configure the Backoffice application** because the ACLs are added during the creation of new pages

---

> id: backoffice-auth0-integration

Reviewer 2: ```suggestion
id: auth0-integration
```

> title: Backoffice Integration with Auth0

Reviewer 2: ```suggestion
title: Integration with Auth0
```

Author: Fixed

sidebar_label: Backoffice Integration with Auth0
---

# Mia-Platform Backoffice Integration with Auth0

In this tutorial, we will guide you through the process of securing Mia-Platform Backoffice using Auth0 as your Identity Provider. In particular, we will see how to:

> - protect the Mia-Platform backoffice with the Auth0 login page;

Reviewer: ```suggestion
- protect the Mia-Platform Backoffice with the Auth0 login page;
```

> - create, delete and manage both Auth0 users and users' groups using the Backoffice;

Reviewer: ```suggestion
- create, delete and manage both Auth0 users and users' groups from dedicated pages on Mia-Platform Backoffice;
```

> - use backoffice ACLs to display to the user only the subset of pages and components they are authorized to access.

Reviewer: ```suggestion
- use the ACLs feature of Mia-Platform Backoffice to display to the user only the subset of pages and components they are authorized to access.
```
